### PR TITLE
Improve the performance of column `scatter`

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...
Improve the performance of column `scatter` by writing into columns sequentially, without switching destination columns frequently which reduce cache miss.

we compare with the previous implementation as following.

query
```sql
-- column scatter is used in parallel_hash
SELECT
    t1.msgtype,
    count(*)
FROM
(
    SELECT
        rip,
        rtime,
        msgid,
        msgtype
    FROM test_data
    WHERE (day = '2023-03-06') AND (hour = '00')
) AS t1
ANY INNER JOIN
(
    SELECT
        rip,
        rtime,
        msgid,
        msgtype
    FROM test_data
    WHERE (day = '2023-03-06') AND (hour = '00')
) AS t2 ON (t1.rip = t2.rip) AND (t1.msgid = t2.msgid) AND (t1.rtime = t2.rtime) AND (t1.msgtype = t2.msgtype)
GROUP BY t1.msgtype
SETTINGS max_threads = 32, join_algorithm = 'parallel_hash'
```

result

|  threads   | before  | after |
|  ----  | ----  | ----  |
| 1 | 6 rows in set. Elapsed: 46.652 sec. Processed 37.40 million rows, 4.53 GB (801.77 thousand rows/s., 97.12 MB/s.) | 6 rows in set. Elapsed: 45.250 sec. Processed 37.40 million rows, 4.53 GB (826.61 thousand rows/s., 100.13 MB/s.) |
| 2  | 6 rows in set. Elapsed: 30.992 sec. Processed 37.40 million rows, 4.53 GB (1.21 million rows/s., 146.19 MB/s.) | 6 rows in set. Elapsed: 26.425 sec. Processed 37.40 million rows, 4.53 GB (1.42 million rows/s., 171.46 MB/s.) |
| 4 | 6 rows in set. Elapsed: 21.150 sec. Processed 37.40 million rows, 4.53 GB (1.77 million rows/s., 214.23 MB/s.) | 6 rows in set. Elapsed: 18.612 sec. Processed 37.40 million rows, 4.53 GB (2.01 million rows/s., 243.43 MB/s.) |
| 8  | 6 rows in set. Elapsed: 18.219 sec. Processed 37.40 million rows, 4.53 GB (2.05 million rows/s., 248.69 MB/s.) | 6 rows in set. Elapsed: 16.080 sec. Processed 37.40 million rows, 4.53 GB (2.33 million rows/s., 281.76 MB/s.) |
| 16 | 6 rows in set. Elapsed: 11.162 sec. Processed 37.40 million rows, 4.53 GB (3.35 million rows/s., 405.92 MB/s.) | 6 rows in set. Elapsed: 11.620 sec. Processed 37.40 million rows, 4.53 GB (3.22 million rows/s., 389.92 MB/s.) |
| 32  | 6 rows in set. Elapsed: 12.200 sec. Processed 37.40 million rows, 4.53 GB (3.07 million rows/s., 371.39 MB/s.) | 6 rows in set. Elapsed: 11.879 sec. Processed 37.40 million rows, 4.53 GB (3.15 million rows/s., 381.41 MB/s.) |

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by vdimir** [comment]
No change for `parallel_hash_join` in https://s3.amazonaws.com/clickhouse-test-reports/47329/8aa18ca1406abc106de682268193d453266866f2/performance_comparison_[2/4]/report.html 🤔 

And could you please check slow downs reported in the reports, can they be related?

**Comment by lgbo-ustc** [comment]
> No change for `parallel_hash_join` in [https://s3.amazonaws.com/clickhouse-test-reports/47329/8aa18ca1406abc106de682268193d453266866f2/performance_comparison_[2/4]/report.html](https://s3.amazonaws.com/clickhouse-test-reports/47329/8aa18ca1406abc106de682268193d453266866f2/performance_comparison_%5B2/4%5D/report.html) 🤔
> 
> And could you please check slow downs reported in the reports, can they be related?

It may be only impove when scatter for blocks. If there is only one column, the cost for building the partition info is large.

**Comment by lgbo-ustc** [comment]
It's weird that the performance test dosn't improve, we will try more tests to see what happen

File: src/Columns/ColumnAggregateFunction.cpp [hunk 1]
```diff
@@ -606,18 +606,24 @@ MutableColumns ColumnAggregateFunction::scatter(IColumn::ColumnIndex num_columns
...
-                column->reserve(reserve_size);
+        auto from = partition_info.partitions_offset[i];
+        auto end = partition_info.partitions_offset[i + 1];
+        if (from == end) [[unlikely]]
+            continue;
+        auto & column = columns[i];
+        column->reserve(end - from);
+        for (size_t j = from; j < end; ++j)
+        {
+            assert_cast<ColumnAggregateFunction &>(*column).data.push_back(data[offsets_map[j]]);
```
Comment (zhanglistar) [inline src/Columns/ColumnAggregateFunction.cpp:624]: prefetch data[offsets_map[j+1]+16]

**Comment by lgbo-ustc** [comment]
With small source size, this implementation will be better.
Maybe we try it later.

**Pull request closed without merging**